via/services/google_drive.py

```diff
@@ -16,29 +17,48 @@
 LOG = getLogger(__name__)
 
 
-def translate_google_error(
-    error,
-):  # pylint:disable=too-many-return-statements
+class GoogleDriveErrorSchema(Schema):
+    """Schema for the JSON bodies of Google Drive error responses."""
+
+    class Meta:
+        unknown = INCLUDE
+
+    class Errors(Schema):
+        class Error(Schema):
+            message = fields.String()
+            reason = fields.String()
+
+        errors = fields.List(
+            fields.Nested(Error(unknown=INCLUDE)),
+            required=True,
+            validate=validate.Length(min=1),
+        )
+
+    @pre_load
+    def pre_load(self, data, **_kwargs):  # pylint:disable=no-self-use
+        try:
+            return data.response.json()
+        except JSONDecodeError as err:
+            raise ValidationError("Invalid JSON") from err
+
+    error = fields.Nested(Errors(unknown=INCLUDE), required=True)
+
+
+def translate_google_error(error):
     """Get a specific error instance from the provided error or None."""
 
-    # This isn't a requests exception we can get meaningful data from
     if not isinstance(error, HTTPError) or error.response is None:
         return None
 
     status_code = error.response.status_code
 
-    # Try and parse out the Google details in the format we've seen
     try:
-        google_error = error.response.json()["error"]["errors"][0]
-    except (JSONDecodeError, KeyError, TypeError):
+        validated_data = GoogleDriveErrorSchema().load(error)
+    except ValidationError:
         return None
```

Comment:
A companion of the above, but I think it's better to keep getting the data outside of the schema. This keeps us less tightly bound to Marshmallow. Also I don't think Marshmallow is actually is offering any additional functionality. You _can_ attach this to the schema, but you can also just not do that quite trivially, and I think it's simpler:

From:

```python
    @pre_load
    def pre_load(self, data, **_kwargs):  # pylint:disable=no-self-use
        try:
            return data.response.json()
        except JSONDecodeError as err:
            raise ValidationError("Invalid JSON") from err     
    
    # meanwhile... in the view code...
    
    try
      validated_data = GoogleDriveErrorSchema().load(error)
    except ValidationError:
        return None
```

To:

```python
try:
    data = request.json()
    GoogleDriveErrorSchema().validate(data)
except ValidationError, JSONDecodeError:
   return None
```

You could split the above up into separate try's if you preferred.

I think there are some other advantages to this:

* The bonus to readability (where does this data come from?) mentioned above
* It's actually a lot shorter
* The schema can trivially be applied to validating things which aren't from a request
* Testing becomes easier:
   * You don't have to worry `validate()`'s return value
   * You don't have to construct a request object, you can just provide a dict
   
There was some discussion of re-use around the `try ... catch JSONDecodeError` which I have a few thoughts on:

* I'm not sure something this basic is actually worth re-using at all. Repetition might be better
* In most views, why catch it at all? This could be dealt with by exception views
* If this is something worth re-using it doesn't necessarily imply it's anything to do with the schema. This looks like it could be a natural fit for a request method?

Reply:
Another suggestion is to not use `GoogleDriveErrorSchema().load()`'s return value since we know it just returns the given data unmodified. This arguably makes it more explicit in the service code that we're just using `GoogleDriveErrorSchema` to validate the data and not to mutate/deserialize it. This is independent of changing from `load()` to `validate()`:

You can't do this:

```python
try:
    json_data = error.response.json()
    GoogleDriveErrorSchema().load(json_data)
except (JSONDecodeError, ValidationError):
    return None
```

because it will trigger PyLint's `try clause contains 2 statements, expected at most 1 (too-many-try-statements)`.

So instead you have to do this:

```python
try:
    json_data = error.response.json()
except JSONDecodeError:
    return None

try:
    GoogleDriveErrorSchema().load(json_data)
except ValidationError:
    return None
```

But this trigger's PyLint's `Too many return statements (7/6) (too-many-return-statements)`. It's also quite a few more lines of code than the current.

Changing from `load()` to `validate()` actually saves a couple of trivial lines of code by replacing a `try` / `except` with an `if`:

```python
try:
    json_data = error.response.json()
except JSONDecodeError:
    return None

if GoogleDriveErrorSchema().validate(json_data):
    return None
```

But it's still longer and still triggering PyLint's `too-many-return-statements`.